lms/djangoapps/support/views/course_entitlements.py

```diff
@@ -0,0 +1,133 @@
+from django.db.models import Q
+from django.http import HttpResponseBadRequest
+from django.utils.decorators import method_decorator
+from edx_rest_framework_extensions.authentication import JwtAuthentication
+from rest_framework import permissions, status, viewsets
+from rest_framework.response import Response
+from six import text_type
+
+from entitlements.api.v1.permissions import IsAdminOrAuthenticatedReadOnly
+from entitlements.api.v1.serializers import SupportCourseEntitlementSerializer
+from entitlements.models import CourseEntitlement, CourseEntitlementSupportDetail
+from openedx.core.djangoapps.cors_csrf.authentication import SessionAuthenticationCrossDomainCsrf
+from lms.djangoapps.support.decorators import require_support_permission
+
+
+class EntitlementSupportView(viewsets.ModelViewSet):
+    """
+    Allows viewing and changing learner course entitlements, used the support team.
+    """
+    authentication_classes = (JwtAuthentication, SessionAuthenticationCrossDomainCsrf,)
+    permission_classes = (permissions.IsAuthenticated, IsAdminOrAuthenticatedReadOnly,)
+    queryset = CourseEntitlement.objects.all()
+    serializer_class = SupportCourseEntitlementSerializer
+
+    def filter_queryset(self, queryset):
+        try:
+            username_or_email = self.request.GET.get('user')
+            user = User.objects.get(Q(username=username_or_email) | Q(email=username_or_email))
+            queryset = queryset.filter(user=user)
+            return super(EntitlementSupportView, self).filter_queryset(queryset).order_by('created')
+        except (KeyError, User.DoesNotExist):
+            return queryset.order_by('created')
+
+    @method_decorator(require_support_permission)
+    def update(self, request):
+        """ Allows support staff to unexpire a user's entitlement."""
+        support_user = request.user
+        try:
+            username_or_email = self.request.data['user']
+            user = User.objects.get(Q(username=username_or_email) | Q(email=username_or_email))
+        except User.DoesNotExist:
+            return HttpResponseBadRequest(
+                u'Could not find user {user}'.format(user=username_or_email)
+            )
+        try:
+            course_uuid = request.data['course_uuid']
+            reason = request.data['reason']
+            comments = request.data.get('comments', None)
+        except KeyError as err:
+            return HttpResponseBadRequest(u'The field {} is required.'.format(text_type(err)))
+        try:
+            entitlement = CourseEntitlement.get_most_recent_entitlement(user=user, course_uuid=course_uuid)
+        except CourseEntitlement.DoesNotExist:
+            return HttpResponseBadRequest(
+                u'Could not find entitlement to course {course_uuid} for user {user}'.format(
+                    course_uuid=course_uuid, user=username_or_email
+                )
+            )
+        if entitlement:
+            if entitlement.expired_at is None:
```

Comment:
I'm trying to think if there are use-cases where a learner has an unexpired entitlement and we want support to be able to modify it. It might not be something we need to support right now, but building this check for entitlement expiration into the update api is locking us into only enabling support to modify expired entitlements, which seems a bit over-protective to me.